clients/src/main/java/org/apache/kafka/clients/consumer/internals/PartitionAssignor.java

```diff
@@ -58,10 +58,21 @@
      */
     Map<String, Assignment> assign(Cluster metadata, Map<String, Subscription> subscriptions);
 
+    /**
+     * Perform the group assignment given the member subscriptions, current cluster metadata and group generation.
+     * @param metadata Current topic/broker metadata known by consumer
+     * @param subscriptions Subscriptions from all members provided through {@link #subscription(Set)}
+     * @param generation Current group generation id
+     * @return A map from the members to their respective assignment. This should have one entry
+     *         for all members who in the input subscription map.
+     */
+    default Map<String, Assignment> assign(Cluster metadata, Map<String, Subscription> subscriptions, int generation) {
```

Comment:
Hmm.. Now that I'm thinking about it, this seems like the wrong place to pass through generation. We do not need to know the current generation in order to sort the current and previous assignments. We should only need to know the current generation when we receive the assignment after the rebalance completes. I think we should have the following method instead:

```java
void onAssignment(Assignment assignment, Optional<Integer> generation);
```

Perhaps this is the cause of the confusion over propagation of the generation through the Assignment UserData.

Reply:
Hmm, would this work? The reason we introduced the `generation` argument in `assign` method was to pass on the generation value from the coordinator to the sticky assignor so we can perform the assignment by comparing the consumer's reported generation/assignment with the current assignment. Passing it on in `onAssignment` only solves one half of the problem (consumers knowing what their assignment generation is). It, however, doesn't help the assignor know the current assignment. Are you implying we don't care what the current assignment is when performing the sticky assignment and just rely on the sorted generation assignments we built?

Reply:
Yes, I'm suggesting that the current generation is not needed when performing the assignment. All we need to know is the generation of the previous assignment reported by each member, which is included in the subscription user data. Does that make sense? Then we get all the information we need for the next rebalance inside `onAssignment`.

Reply:
Ok. Thanks for clarifying. That should work. I'll work on modifying the code accordingly later today.